### PR TITLE
Autogenerate index

### LINKED ISSUE #2094: [TASK] Ability to automatically generate an index
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
Some times a code will out put a csv file that doesn't have a useful column that can be used as a dataset index. It would be useful if this could automatically be generated.

**Describe the solution you'd like**
Add something that allows a code or a dataset to have an automatically generated index.

**Describe alternatives you've considered**
Make a custom Code Interface just to create an index column.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by joshua-cogliati-inl**
--------
Pull Request Description
--------
##### What issue does this change request address?
Closes #2094 

##### What are the significant changes in functionality due to this change request?
Adds the ability for a DataSet to have an automatically generated index.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: doc/user_manual/database_data.tex [hunk 1]
```diff
@@ -148,6 +148,10 @@ \section{DataObjects}
    \begin{itemize}
      \item \xmlAttr{var}, \xmlDesc{required string attribute}, the dimension name of this index (e.g. time)
    \end{itemize}
+   This XML node allows the following attribute:
+   \begin{itemize}
+     \item \xmlAttr{autogenerate}, \xmlDesc{optional boolean attribute}, if this index should be generated automatically if it does not exist.
```
Comment (wangcj05) [inline doc/user_manual/database_data.tex:153]: Could you add more detail on this feature?

File: ravenframework/DataObjects/DataSet.py [hunk 2]
```diff
@@ -211,6 +231,13 @@ def addRealization(self, rlz):
     if indexMap is not None:
       # keep only those parts of the indexMap that correspond to variables we care about.
       indexMap = dict((key, val) for key, val in indexMap[0].items() if key in self.getVars()) # [0] because everything is nested in a list by now, it seems
+    #If the index is in autogenerate set, generate it automatically
+    if len(self._autogenerate) > 0:
+      for autoindex in self._autogenerate:
+        expectedLength = len(rlz[self._pivotParams[autoindex][0]])
+        #if it already exists and has correct length, don't add
+        if autoindex not in rlz or len(rlz[autoindex]) != expectedLength:
+          rlz[autoindex] = np.arange(len(rlz[self._pivotParams[autoindex][0]]))
```
Comment (wangcj05) [inline ravenframework/DataObjects/DataSet.py:240]: change len(rlz[self._pivotParams[autoindex][0]]) to expectedLength?

File: tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv [hunk 3]
```diff
@@ -0,0 +1,5 @@
+autoindex,power,x,y,z,temperature
+0,100.0,1.0,1.0,1.0,3.0
+1,100.0,2.0,1.0,1.0,5.0
+0,200.0,1.0,1.0,1.0,8.0
+1,200.0,2.0,1.0,1.0,10.0
```
Comment (wangcj05) [inline tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv:5]: From this results, it seems to me the autoindex  is used to label the index in each sampled output, i.e., the index for each row in sample/1/All.csv. If this is the case, it seems it is inconsistent for us to treat the samples. For raven, each sample file is one realization, but in this case, each sample file contains multiple realizations. Could you comment on this?

File: tests/framework/CodeInterfaceTests/test_generic_dataset.xml [hunk 4]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" ?>
+<Simulation verbosity="debug">
```
Comment (wangcj05) [inline tests/framework/CodeInterfaceTests/test_generic_dataset.xml:2]: Please add TestInfo block.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@joshua-cogliati-inl I have some general comments for you to consider.

File: tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv (hunk 3 above)
Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv:5]: This is useful for supporting cases where each sample outputs multiple data points, such as a temperature for each x,y,z location.

File: tests/framework/CodeInterfaceTests/test_generic_dataset.xml (hunk 4 above)
Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/test_generic_dataset.xml:2]: Fixed.

File: ravenframework/DataObjects/DataSet.py (hunk 2 above)
Comment (joshua-cogliati-inl) [reply ravenframework/DataObjects/DataSet.py]: I agree. fixed.

File: doc/user_manual/database_data.tex (hunk 1 above)
Comment (joshua-cogliati-inl) [reply doc/user_manual/database_data.tex]: Added more detail.

File: tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv (hunk 3 above)
Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv:5]: I'm thinking to have two separate indices, one is the "RAVEN_Sample_ID", which is automatically generated, and the other is autoindex. From this gold file, it seems only autoindex is used. I would like to add  "RAVEN_Sample_ID" (or prefix). In this, we will have two indices that uniquely determine each row of data. Please correct me if I'm wrong.

Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv:5]: I would suggest we add it.

Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv:5]: Actually, we do have an automatically generated sample index already.  It is called 'CURRENT_ID' so if the dataset is:
```xml
    <DataSet name="samples">
      <Input>power, autoindex</Input>
      <Output>x, y, z, temperature, CURRENT_ID</Output>
      <Index var="autoindex" autogenerate="True">x, y, z, temperature</Index>
    </DataSet>
```
we get this csv:
```
autoindex,power,x,y,z,temperature,CURRENT_ID
0,100.0,1.0,1.0,1.0,3.0,0
1,100.0,2.0,1.0,1.0,5.0,0
0,200.0,1.0,1.0,1.0,8.0,1
1,200.0,2.0,1.0,1.0,10.0,1
```
This is sort of an undocumented feature (we use CURRENT_ID internally)
Should this be documented? Should I add this to the test?

Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv:5]: @wangcj05 We have RAVEN_sample_ID in various places, but that currently is not accessible when creating the dataset (without code additions).  prefix also works and might be better than CURRENT_ID.  Let me know how you would like to solve this.

Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv:5]: I prefere prefix since it is used more often than CURRENT_ID.

Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv]: I added prefix to the input file and the gold file.

Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv]: Thanks a lot @joshua-cogliati-inl. Should we update the manual to explain it? what do you think?

Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv]: I am not sure.  We could probably mention it in the data objects part of the manual.

File: ravenframework/DataObjects/DataSet.py [hunk 5]
```diff
@@ -65,6 +65,25 @@ def __init__(self):
...
+  @classmethod
+  def getInputSpecification(cls):
+    """
+      Get a reference to a class that specifies the input data for class "cls".
+      @ In, cls, the class for which we are retrieving the specification
+      @ Out, inputSpecification, InputData.ParameterInput, class to use for specifying the input of cls.
+    """
+    inputSpecification = super(DataSet,cls).getInputSpecification()
+
+    # TODO this should be specific to ND set
```
Comment (joshua-cogliati-inl) [inline ravenframework/DataObjects/DataSet.py:80]: This TODO has been fixed, so remove TODO from it.

File: tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv (hunk 3 above)
Comment (joshua-cogliati-inl) [reply tests/framework/CodeInterfaceTests/gold/GenericDataSet/samples.csv]: I mentioned prefix in the manual.

File: ravenframework/DataObjects/DataSet.py (hunk 5 above)
Comment (joshua-cogliati-inl) [reply ravenframework/DataObjects/DataSet.py]: Fixed.

**Review by wangcj05 — APPROVED** [review]
PR looks good.

**Comment by wangcj05** [comment]
checklist is good.

**Pull request merged**